src/main/java/hudson/remoting/ClassFilter.java

```diff
@@ -174,31 +197,36 @@ public final Class check(Class c) {
          */
         private static final Pattern OPTIMIZE2 = Pattern.compile("^\\^\\Q[^\\\\]+\\\\E\\.\\*$");
 
-        private final Object[] blacklistPatterns;
+        private final List<Object> blacklistPatterns;
 
-        public RegExpClassFilter(List<Pattern> blacklistPatterns) {
-            this.blacklistPatterns = blacklistPatterns.toArray(new Pattern[blacklistPatterns.size()]);
+        RegExpClassFilter(String[] patterns) throws ClassFilterException {
+            blacklistPatterns = new ArrayList<>(patterns.length);
+            for (String pattern : patterns) {
+                add(pattern);
+            }
         }
 
-        RegExpClassFilter(String[] patterns) {
-            blacklistPatterns = new Object[patterns.length];
-            for (int i = 0, patternsLength = patterns.length; i < patternsLength; i++) {
-                if (OPTIMIZE1.matcher(patterns[i]).matches()) {
-                    // this is a simple startsWith test, no need to slow things down with a regex
-                    blacklistPatterns[i] = patterns[i].substring(1,patterns[i].length()-2).replace("[.]",".");
-                } else  if (OPTIMIZE2.matcher(patterns[i]).matches()) {
-                    // this is a simple startsWith test, no need to slow things down with a regex
-                    blacklistPatterns[i] = patterns[i].substring(3,patterns[i].length()-4);
-                } else {
-                    blacklistPatterns[i] = Pattern.compile(patterns[i]);
+        void add(String pattern) throws ClassFilterException {
+
+            if (OPTIMIZE1.matcher(pattern).matches()) {
+                // this is a simple startsWith test, no need to slow things down with a regex
+                blacklistPatterns.add(pattern.substring(1, pattern.length() - 2).replace("[.]", "."));
+            } else if (OPTIMIZE2.matcher(pattern).matches()) {
+                // this is a simple startsWith test, no need to slow things down with a regex
+                blacklistPatterns.add(pattern.substring(3, pattern.length() - 4));
```

Comment:
Perhaps it also makes sense to handle out of the bounds exception in this PR. Not a new issue though